plugins/embed-optimizer/hooks.php

```diff
@@ -23,90 +53,156 @@
+	$bookmark_names = array(
+		'script' => 'embed_optimizer_script',
+		'iframe' => 'embed_optimizer_iframe',
+	);
+	try {
+		/*
+		 * Determine how to lazy load the embed.
+		 *
+		 * - If there is only one iframe, set loading="lazy".
+		 * - Prevent making scripts lazy if there is an inline script.
+		 * - Only make script lazy if there is a single external script (since if there are
+		 *   multiple they may not get loaded in the right order).
+		 * - Ensure that both the iframe and the script are made lazy if both occur in the same embed.
+		 */
+		$iframe_count      = 0;
+		$script_count      = 0;
+		$needs_lazy_script = false;
+		$has_inline_script = false;
+		$figure_depth      = 0;
+		// Locate the iframes and scripts.
+		do {
+			// This condition ensures that when iterating over an embed inside a larger document that we stop once we reach
+			// closing </figure> tag. The $processor is an OD_HTML_Tag_Processor when Optimization Detective is iterating
+			// over all tags in the document, and this embed_optimizer_update_markup() is usd as part of the tag visitor
+			// from Embed Optimizer. On the other hand, if $html_processor is not an OD_HTML_Tag_Processor then this is
+			// iterating over the tags of the embed markup alone as is passed into the embed_oembed_html filter.
+			if ( $html_processor instanceof OD_HTML_Tag_Processor ) {
+				if ( 'FIGURE' === $html_processor->get_tag() ) {
+					if ( $html_processor->is_tag_closer() ) {
+						--$figure_depth;
+						if ( $figure_depth <= 0 ) {
+							// We reached the end of the embed.
+							break;
+						}
+					} else {
+						++$figure_depth;
+						// Move to next element to start looking for IFRAME or SCRIPT tag.
+						continue;
+					}
+				}
+				if ( 0 === $figure_depth ) {
+					continue;
 				}
 			}
-		} elseif ( 'SCRIPT' === $html_processor->get_tag() ) {
-			if ( ! $html_processor->get_attribute( 'src' ) ) {
-				$has_inline_script = true;
-			} else {
-				++$script_count;
-				if ( ! $html_processor->set_bookmark( 'script' ) ) {
-					wp_trigger_error( __FUNCTION__, esc_html__( 'Embed Optimizer unable to set script bookmark.', 'embed-optimizer' ) );
-					return $html;
+
+			if ( 'IFRAME' === $html_processor->get_tag() ) {
+				$loading_value = $html_processor->get_attribute( 'loading' );
+				if ( empty( $loading_value ) ) {
+					++$iframe_count;
+					if ( ! $html_processor->set_bookmark( $bookmark_names['iframe'] ) ) {
+						throw new Exception( __( 'Embed Optimizer unable to set iframe bookmark.', 'embed-optimizer' ) );
+					}
+				}
+			} elseif ( 'SCRIPT' === $html_processor->get_tag() ) {
+				if ( ! $html_processor->get_attribute( 'src' ) ) {
+					$has_inline_script = true;
+				} else {
+					++$script_count;
+					if ( ! $html_processor->set_bookmark( $bookmark_names['script'] ) ) {
+						throw new Exception( __( 'Embed Optimizer unable to set script bookmark.', 'embed-optimizer' ) );
```

Comment:
Use sprintf placeholders for `script` and `iframe` here so that this becomes the same string for both.